### PR TITLE
Dual stack probes limitations

### PR DESCRIPTION
**Pull request by maiqueb**
Document limitations of HTTP / TCP probes on dual stack networking configurations.

### CODE REVIEW

File: creation/probes.md [hunk 1]
```diff
@@ -169,3 +169,30 @@ filled:
...
+(or
+[TCP](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#tcpsocketaction-v1-core))
+probe actions `host` optional field default to the pod's IP address - which, is
+currently **always** the IPv4 address.
+
+Since the probe must be specified when defining the VMI, it is impossible to
+fill the `host` field with the IP address - we do not have the pod IP address
+yet.
+
+A work-around for this would be to expose the VMI server replying to the probe
```
Comment (RamLavi) [inline creation/probes.md:197]: I think this last line would benefit by an example, or a link to somewhere it is done like so.

Comment (maiqueb) [reply creation/probes.md:197]: I get the feeling that's a bit besides the point ... What I could do is refer the users to https://github.com/kubevirt/user-guide/blob/master/usage/network-service-integration.md#expose-virtualmachineinstance-as-a-clusterip-service , which explains how to do that . 

It would probably be safer to just remove the whole paragraph, since I have *never* tried it (I just don't see why it wouldn't work) ...

I'll make sure it works, then either refer the users to the service API, *or* get rid of this.

File: creation/probes.md [hunk 2]
```diff
@@ -63,7 +63,7 @@ port 1500 via cloud-init.
             bootcmd:
               - setenforce 0
               - dnf install -y nmap-ncat
-              - systemd-run --unit=httpserver nc -klp 1500 -e '/usr/bin/echo -e HTTP/1.1 200 OK\\n\\nHello World!'
+              - systemd-run --unit=httpserver nc -klp 1500 -e '/usr/bin/echo -e HTTP/1.1 200 OK\\nContent-Length: 12\\n\\nHello World!'
```
Comment (RamLavi) [inline creation/probes.md:66]: I'm guessing this is the same fix as this [PR](https://github.com/kubevirt/kubevirt/pull/3741)
can you add a clarification on the commit message?

File: creation/probes.md [hunk 3]
```diff
@@ -169,3 +169,30 @@ filled:
...
+the VM.
+
+Dual-stack networking configurations have a limitation when using HTTP / TCP
+probes - you **cannot probe the VMI by its IPv6 address**. The reason for this
+is the
+[HTTP](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#httpgetaction-v1-core)
+(or
+[TCP](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#tcpsocketaction-v1-core))
+probe actions `host` optional field default to the pod's IP address - which, is
+currently **always** the IPv4 address.
```
Comment (oshoval) [inline creation/probes.md:191]: want to add 
https://github.com/containerd/cri/issues/1278 here ?

Comment (maiqueb) [reply creation/probes.md:191]: Not sure that's valuable information for a person attempting to configure a probe; what the user is after is how to configure the feature. 

What I could do instead is link that in the commit msg, so 'we' can track that issue and update the user guide, thus giving the user a valid, up-to-date picture.

File: creation/probes.md (hunk 2 above)
Comment (maiqueb) [reply creation/probes.md:66]: This is about providing a correct example to the user.

Sure, I can link that PR in the commit message, for completeness.

Comment (maiqueb) [reply creation/probes.md:66]: Please resolve if you think the commit msg features all the info required.

File: creation/probes.md (hunk 3 above)
Comment (maiqueb) [reply creation/probes.md:191]: Done; please resolve if you think the commit msg features all the info required.

File: creation/probes.md (hunk 2 above)
Comment (RamLavi) [reply creation/probes.md:66]: I don't have a resolve button, but consider this resolved :)

File: creation/probes.md (hunk 3 above)
Comment (oshoval) [reply creation/probes.md:191]: (reviewer cant resolve)

File: creation/probes.md (hunk 1 above)
Comment (maiqueb) [reply creation/probes.md]: Despite having the DNS resolution correctly evaluated, I couldn't get a liveness probe to work using the work-around.

As such, since I don't have time to debug the scenario, I prefer to just remove the section about the work-around.

**Comment by maiqueb** [comment]
> /lgtm

I'm afraid this repo doesn't have the bot around to do our bidding. You need to use good old 'approve' button.

**Review by RamLavi — DISMISSED** [review]
Lgtmazing

File: creation/probes.md [hunk 4]
```diff
@@ -169,3 +169,27 @@ filled:
...
+Some context is needed to understand the limitations imposed by a dual-stack
+network configuration on readiness - or liveness - probes. Users must be
+fully aware that a dual-stack configuration is currently only available when
+using a masquerade binding type. Furthermore, it must be recalled that
+accessing a VM using masquerade binding type is performed via the pod IP
+address; in dual-stack mode, both IPv4 and IPv6 addresses can be used to reach
+the VM.
+
+Dual-stack networking configurations have a limitation when using HTTP / TCP
+probes - you **cannot probe the VMI by its IPv6 address**. The reason for this
```
Comment (davidvossel) [inline creation/probes.md:185]: i might rephrase this a bit to read more like

"The reason for this is the host field for both the HTTP and TCP probe actions default to the pod's IP address, which is currently always the IPv4 address.

File: creation/probes.md [hunk 5]
```diff
@@ -169,3 +169,27 @@ filled:
...
+is the
+[HTTP](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#httpgetaction-v1-core)
+(or
+[TCP](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#tcpsocketaction-v1-core))
+probe actions `host` optional field default to the pod's IP address - which, is
+currently **always** the IPv4 address.
+
+Since the probe must be specified when defining the VMI, it is impossible to
```
Comment (davidvossel) [inline creation/probes.md:193]: maybe reword this a bit as well. something like

"Since the pod's IP address is not known before creating the VMI, it isn't possible to pre-populate the probe's `host` field."

File: creation/probes.md (hunk 4 above)
Comment (maiqueb) [reply creation/probes.md:185]: Done.

File: creation/probes.md (hunk 5 above)
Comment (maiqueb) [reply creation/probes.md]: Done.

**Review by RamLavi — APPROVED** [review]
not sure my approval was set (I did it via the app) 
anyway, lgtm

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

Looks great, I have no idea if i have review/approve privileges, but i'll try :D
